Prevent strong chance of failed physical point inversion

lindsayad opened the pull request:
Right now when someone calls `FEProblemBase::reinitElemPhys` with a
vector of physical points, we will attempt to invert those physical
points into the reference element space for both undisplaced and
displaced meshes if the latter exists. It is not difficult to imagine
that for a given set of physical points you may easily fall off the
displaced mesh for large enough displacements/deformations. Let's see if
this makes anything fail

Refs #10823

lindsayad committed:
Prevent strong chance of failed physical point inversion

Right now when someone calls `FEProblemBase::reinitElemPhys` with a
vector of physical points, we will attempt to invert those physical
points into the reference element space for both undisplaced and
displaced meshes if the latter exists. It is not difficult to imagine
that for a given set of physical points you may easily fall off the
displaced mesh for large enough displacements/deformations. The same
issues exists for `FEProblemBase::reinitNeighborPhys`. Callers of these
phys methods should call through the appropriate `SubProblem` that is
relevant to their context. Moreover, we should always make sure that the
caller is calling with an element corresponding to the `SubProblem`
they've selected.

Closes #17355

tophmatthews commented:
Need to comment out `suppress_displaced_init` in the call

lindsayad commented:
If this `reinitElemPhys` patch passes all tests maybe we can wait to do `reinitNeighborPhys` in a later PR. Does that sound good @tophmatthews ?

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1735,7 +1735,7 @@ void
 FEProblemBase::reinitElemPhys(const Elem * elem,
                               const std::vector<Point> & phys_points_in_elem,
                               THREAD_ID tid,
-                              bool suppress_displaced_init)
```
This isn't used in the DisplacedProblem method either - can we remove it? Or are there other applications that still pass something in here?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/707789/) on 620fec6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17355/site)

This comment will be updated on new commits.

tophmatthews commented:
Whats up with this?

lindsayad commented:
Divided human resource :smile:

lindsayad commented:
Trying to get it through though! Made some progress...

lindsayad commented:
This is a nightmare. A total nightmare. The `Constraint` system and its associated logic in `NonlinearSystemBase` never ceases to horrify me. It's the single biggest time-suck trying to sort through these problems. Never-ending streams of coding without encapsulating. Always going through `_fe_problem`.... Aaaaaaaaaaaaaaaaaaaarrrrrrrrrrrrrrrrrrrgggggggggggggggghhhhhhhhhhhhhhhhh

lindsayad committed:
Clear all dof indices at the beginning of a solve

This prevents bad dof indices from lingering around between
invocations of `FEProblemBase::solve`. Here "bad" dof indices
could mean dof indices that no longer exist due to mesh
adaptivity

lindsayad commented:
Oh it's looking like a great success!

lindsayad commented:
@tophmatthews have you tried this on your problematic case?

lindsayad commented:
If it works, can you review/approve this?

tophmatthews commented:
> If it works, can you review/approve this?

I've been staying away due to all the red in civet, but it looks good now, let me test it.

lindsayad commented:
@loganharbour unless @tophmatthews comes back with bad news, I think this is ready to review/go

tophmatthews commented:
Ugg...I'm trying to reproduce...taking longer than I thought...

loganharbour commented on framework/src/problems/DisplacedProblem.C:
```diff
@@ -617,9 +617,11 @@ DisplacedProblem::reinitElem(const Elem * elem, THREAD_ID tid)
-                                 bool)
+                                 THREAD_ID tid)
 {
+  mooseAssert(_mesh.getMesh().query_elem_ptr(elem->id()) == elem,
+              "Are you calling this method with a undisplaced mesh element?");
```
Beautiful assertions 🙌🏼 

FYI in the future for even less code, we have the `MooseMesh::queryElemPtr()` API

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5002,6 +4992,10 @@ FEProblemBase::solve()
 {
   TIME_SECTION(_solve_timer);
 
+  clearAllDofIndices();
```
I think the reasoning is clear, but I think it's still a good idea to put a comment here as to why you ended up needing to do this

loganharbour commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -981,6 +981,12 @@ NonlinearSystemBase::setConstraintSecondaryValues(NumericVector<Number> & soluti
   std::map<std::pair<unsigned int, unsigned int>, PenetrationLocator *> * penetration_locators =
       NULL;
 
+  mooseAssert(displaced ? bool(_fe_problem.getDisplacedProblem().get()) : true,
+              "If we're calling this method with displaced = true, then we better well have a "
+              "displaced problem");
```
```suggestion
  if (displaced)
    mooseAssert(_fe_problem.getDisplacedProblem().get(),
               "If we're calling this method with displaced = true, then we better well have a "
               "displaced problem");
```

With the assertion opt being:

```
if (displaced)
  (void)0
```

this is fine and will be removed by the compiler. I used to do this, but the ternary assertion with a true just looks like hot trash IMO

loganharbour commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1115,6 +1121,12 @@ NonlinearSystemBase::constraintResiduals(NumericVector<Number> & residual, bool
   std::map<std::pair<unsigned int, unsigned int>, PenetrationLocator *> * penetration_locators =
       NULL;
 
+  mooseAssert(displaced ? bool(_fe_problem.getDisplacedProblem().get()) : true,
+              "If we're calling this method with displaced = true, then we better well have a "
+              "displaced problem");
```
Same here

loganharbour commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1852,6 +1864,12 @@ NonlinearSystemBase::constraintJacobians(bool displaced)
   std::map<std::pair<unsigned int, unsigned int>, PenetrationLocator *> * penetration_locators =
       NULL;
 
+  mooseAssert(displaced ? bool(_fe_problem.getDisplacedProblem().get()) : true,
+              "If we're calling this method with displaced = true, then we better well have a "
+              "displaced problem");
```
And here

loganharbour commented on framework/src/systems/SystemBase.C:
```diff
@@ -1549,6 +1549,13 @@ SystemBase::jacobianSetup()
+void
+SystemBase::clearAllDofIndices()
+{
+  for (const auto tid : make_range(libMesh::n_threads()))
+    _vars[tid].clearAllDofIndices();
```
```suggestion
  for (auto & var_warehouse : _vars)
    var_warehouse.clearAllDofIndices();
```

Why not?

loganharbour reviewed (changes requested):
All code style requests, I think.

Removing the extra calls from the reinit methods in FEProblemBase makes this read SO much better

lindsayad committed:
Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/712904/) on 849054e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17355/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17355/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4f46dad974776ed8ad65a193e985d076a81e5495`

lindsayad commented:
@tophmatthews how you doing over there?

tophmatthews commented:
I'm trying...I have so many balls in the air, I can't find a reproducible case that fails obviously to really test this. I'll keep looking, but if tests pass then, don't wait for me...

tophmatthews commented:
Wow, it took forever, but yes, can confirm 👍 

You da 💣  @lindsayad
